src/compiler/ruby_generator.cc

```diff
@@ -37,16 +38,61 @@ using std::vector;
 namespace grpc_ruby_generator {
 namespace {
 
+// returns the full canonical message name
+grpc::string GetCanonicalMessageType(const Descriptor* msg) {
+  const Descriptor* top_level_msg;
+  const FileDescriptor* file_containing_msg;
+  grpc::string msg_full_name;
+  grpc::string msg_name;
+  grpc::string msg_proto_pkg;
+  grpc::string resolved_namespace;
+  grpc::string res(msg->full_name());
+
+  // If msg is nested, find the topmost message
+  top_level_msg = msg;
+  while (top_level_msg->containing_type()) {
+    top_level_msg = top_level_msg->containing_type();
+  }
+  file_containing_msg = top_level_msg->file();
+
+  msg_full_name = top_level_msg->full_name();
+  msg_name = top_level_msg->name();
+  msg_proto_pkg =
+      msg_full_name.substr(0, msg_full_name.length() - msg_name.length());
+
+  if (file_containing_msg->options().has_ruby_package()) {
+    resolved_namespace = file_containing_msg->options().ruby_package();
+  } else {
+    return res;
+  }
+
+  // remove trailing period, prevent it from being replaced
+  if (msg_proto_pkg.length() > 0) {
+    if (msg_proto_pkg.back() == '.') {
+      msg_proto_pkg.pop_back();
+    }
+  } else if (resolved_namespace.length() > 0) {
```

Comment:
If `msg_proto_pkg` is not specified, but the resolved namespace of the message is, then we still want the generated output to have the ruby message be prefixed with the `ruby_package`, right?

From what I can tell, that case would result in just returning `msg->full_name()`, because the "from" in the ReplacePrefix below would be zero length. Right?

Reply:
When "from" in ReplacePrefix is the empty string, ReplacePrefix will return "s" with "to" prepended to it.

```c++
bool ReplacePrefix(grpc::string* s, const grpc::string& from,
                          const grpc::string& to) {
```

![image](https://user-images.githubusercontent.com/21096727/51274943-5f327d80-1985-11e9-8158-b0b569495ee7.png)

Reply:
hmm, after thinking about this I think it might be simpler to handle these two cases separately:

what do you think about something like:

```
if (msg_proto_pkg.length() > 0) {
  msg_proto_pkg.pop_back();
  grpc::string res(msg->full_name());
  ReplacePrefix(&res, msg_proto_pkg, resolved_namespace);
  return res;
} else if (resolved_namespace.length() > 0) {
  return resolved_namespace + '.' + msg->full_name();
} else {
  return msg->full_name();
}
```

then we don't need to "overload" `ReplacePrefix`